documentation/benchmark.md

```diff
@@ -0,0 +1,130 @@
+# Benchmark framework
+## Contents
+1. [Introduction](#1-introduction)
+2. [Implementation details](#2-implementation-details)
+	* [Design overview](#21-design-overview)
+	* [Timestamp source](#22-timestamp-source)
+	* [Call sequence diagram](#23-call-sequence-diagram)
+3. [Running benchmarks](#3-running-benchmarks)
+    * [Benchmark application usage](#31-benchmark-application-usage)
+    * [Adding custom timestamps](#32-adding-custom-timestamps)
+    * [Analyzing results](#33-analyzing-results)
+4. [Limitations and further steps](#4-limitations-and-further-steps)
+
+
+## 1. Introduction
+Due to its nature, OP-TEE is being a solution spanning over several
+architectural layers, where each layer includes its own complex parts.
+For further optimizations of performance, there is a need of tool which will
+provide detailed and precise profiling information for each layer.
+
+It is necessary to receive measurements of latency for:
+* The roundtrip time for going from a client application in normal world,
+down to a Trusted Application and back again.
+* Detailed information for amount of time taken to go through each layer:
+	* libTEEC -> OP-TEE KMOD
+	* OP-TEE KMOD -> OP-TEE OS Core
+	* OP-TEE OS Core -> TA entry point (**not supported yet**)
+	* The same way back
+---
+## 2. Implementation details
+### 2.1 Design overview
+
+Benchmark framework consists of such components:
+1. **Benchmark CA**: a dedicated client application, which is responsible
+for allocating timestamp circular buffers, registering these buffers in
+the **Benchmark PTA**, and consuming all timestamp data, generated
+by all OP-TEE layers. Finally, it puts timestamp data into appropriate
+file with `.ts` extension. Additional details can be found here
+here [optee_benchmark]
+2. **Benchmark PTA**: pseudo TA, which owns all per-cpu circular buffers.
+**Benchmark PTA** must be invoked (by a CA) to register the time stamp
+circular buffers. In turn, the pTA invokes the optee linux driver
+(through some RPC mean) to register this circular buffers in the linux
+kernel layer.
+3. **libTEEC** and **Linux OP-TEE kernel driver** include means to store
+timestamps in the circular buffers once these are registered by the
+**Banchmark pTA** into the "Linux OP-TEE kernel driver".
+
+Benchmark is enabled, all OP-TEE layers (libteec, kernel driver, core)
+do fill the registered timestamp circular buffer with timestamp data for
+all invocation requests
```

Comment:
Add terminal dot.

Suggest adding this: `... for all invocation requests on condition that the circular buffer is allocated/registered.`